stream_create: Add "all users: N" pill to input on clicking `Add all users`

shubham-padia opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #30690 . This PR implements the second remaining part of #30690.

Changes in addition to the original issue description:

1. For the user group popover, the `group settings` link is not shown for system groups after this PR since they are immutable.
2. For user group popover, it only showed the direct members of the group before this PR. Now, we should all members of a group including subgroup members. This will change the behaviour everywhere where the user group popover is being used, but since that might the desired behaviour, that shouldn't be an issue hopefully.
3. User group pill did not count subgroup members as part of its count before, now it does. This will affect all user group pills, but since that might be the desired behaviour, that shouldn't be an issue hopefully.

For system groups, the directive is that we should use the description as the display name. But in case of `role:everyone`, the description is `Admins, moderators, members and guests`, while for the `user_group_pill`, we want to display a simpler and succinct message: `Everyone`. We've only hardcoded this for user_group_pill since we don't want to display the name as `Everyone` anywhere else yet.

`Everyone` cannot be searched on the typeahead and only way to add it is to click the `Add all users` button.

On the technical side, please refer to the commit messages and the PR comments for more details.
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<img width="730" alt="Screenshot 2024-07-09 at 7 02 41 PM" src="https://github.com/zulip/zulip/assets/13910561/03fd86c7-a28a-424a-a125-3764a5e66c75">

<details>
<summary> Everyone user group popover (no settings link) </summary>
<img width="461" alt="Screenshot 2024-07-09 at 7 25 28 PM" src="https://github.com/zulip/zulip/assets/13910561/ed543571-0138-408f-942d-8ea32a9a4752">


</details>

<details>
<summary> Non system user group popover (settings link is present) </summary>
<img width="731" alt="Screenshot 2024-07-09 at 7 25 45 PM" src="https://github.com/zulip/zulip/assets/13910561/2a495edd-0251-40df-85e0-31032bdaccbc">


</details>

<details>
<summary> All users popover with deactivated users (12 instead of the usual 16 users) </summary>

<img width="412" alt="Screenshot 2024-07-11 at 4 22 54 PM" src="https://github.com/zulip/zulip/assets/13910561/6c41d9b9-0ecc-4423-b729-51bc89c397eb">


</details>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

shubham-padia commented on web/tests/user_groups.test.js:
```diff
@@ -190,6 +190,66 @@ run_test("get_recursive_subgroups", () => {
+    // to verify whether our algorithm correctly avoids visiting groups multiple times
+    // when determining recursive subgroups.
+    // A test case that can occur in practice and would be problematic without this
+    // optimization is a tree where each layer connects to every node in the next layer.
```
I've kept the test case mentioned in this comment around for this set of tests. 

While `get_recursive_group_members` uses `get_recursive_subgroups` underneath, and this case has been copied from the `get_recursive_subgroups` tests, having this case repeat here again would be good to have if the underlying implementation to get members changes in the future.

shubham-padia commented on web/src/user_groups.ts:
```diff
@@ -202,6 +202,23 @@ export function get_recursive_subgroups(target_user_group: UserGroup): Set<numbe
+
+    if (subgroup_ids === undefined) {
+        return members;
+    }
```
I'm returning the direct members of our `target_user_group` if we have an error while trying to find subgroups. I'm not sure in which circumstances we would end up with undefined in the real world scenario, so I've just decided to return the direct members for now. Let me know if this needs to change

shubham-padia commented on web/src/user_groups.ts:
```diff
@@ -202,6 +202,23 @@ export function get_recursive_subgroups(target_user_group: UserGroup): Set<numbe
+    const subgroup_ids = get_recursive_subgroups(target_user_group);
+
+    if (subgroup_ids === undefined) {
+        return members;
```
I could also add an early return when the subgroup id set size is 0, but there's no benefit to doing that since the for loop below is gonna get skipped anyways in that case and the next statement is `return members`

timabbott replied:
It looks like `get_recursive_subgroups` can only return `undefined` if a subgroup was not defined? I'm not sure that's possible today, but probably this will need to be reworked if we add restrictions on viewing group membership.

shubham-padia commented on web/src/user_group_pill.ts:
```diff
@@ -47,17 +51,18 @@ export function get_group_name_from_item(item: InputPillItem<UserGroupPill>): st
     user_ids = [...new Set(user_ids)];
     user_ids.sort((a, b) => a - b);
 
-    user_ids = user_ids.filter(Boolean);
```
This was introduced in ce4cf66f3f80c3b120507d6907b9197c343f81eb, not sure why. Let me know if there's a reason for this to be here and we can drop the first commit that removes this

timabbott replied:
My guess is it was to deal with some previous construction that would put `undefined` values into the list. I think we can remove it, as no such construction is visible here.

alya commented:
Works for me in manual testing! I'm not sure wether it should be capitalized (maybe group names often won't be?), but I think it looks nice capitalized, so maybe it's good.

alya commented:
Chatting with @timabbott , this part sounds good:

> User group pill did not count subgroup members as part of its count before, now it does. This will affect all user group pills

alya commented:
> For user group popover, it only showed the direct members of the group before this PR. Now, we should all members of a group including subgroup members.

@timabbott and I are currently thinking this makes sense to do when the group you are looking at (the outer one) is a system group (including Everyone, as here). Users don't really need to know how system groups are implemented internally.

When the outer group is not a system group, I think we'll want to list subgroups by name, without exploding them. Would it make sense to adjust this PR accordingly?

shubham-padia commented:
> When the outer group is not a system group, I think we'll want to list subgroups by name, without exploding them. Would it make sense to adjust this PR accordingly?

That would be a good thing to do. But is there a way to create a non-system user group with subgroups in the UI? I tried to create one in Group settings + looked at Zulip help for ways to do it, but couldn't find a way to do it.

If there's no way to create those kinds of non-system user group, we can also add support for the group card when we add a way to add subgroups in UI.

alya commented:
Probably best for @sahil839 to take a look a now and comment on how we should proceed.

sahil839 commented on web/src/add_subscribers_pill.ts:
```diff
@@ -132,6 +132,10 @@ export function create_without_add_button({
     return pill_widget;
 }
 
+export function append_pill_from_text(text: string, pill_widget: CombinedPillContainer): void {
```
Can we use `user_group_pill.append_user_group` here?

shubham-padia replied:
Fixed! Added a function to add a user group pill from group name for now since that is all we need (not stream or user)

sahil839 commented:
I should not be tough to handle non-system groups separately, but we can also leave it for later.

shubham-padia commented:
Opened a discussion on a detail missed before while testing this PR: [link](https://chat.zulip.org/#narrow/stream/101-design/topic/User.20group.20popovers.20-.20deactivated.20users.20-.20stream.20create/near/1883376).

shubham-padia commented:
I've pushed a fix that excludes the inactive users in the last commit and added a screenshot to the PR too. While https://github.com/zulip/zulip/pull/30768#issuecomment-2221926595 can be a potential blocker, the PR should be reviewable while that gets discussed.

shubham-padia commented on web/src/add_subscribers_pill.ts:
```diff
@@ -132,6 +134,21 @@ export function create_without_add_button({
+    if (user_group === undefined) {
+        // This shouldn't happen, but we'll give a warning for now if it
+        // does.
+        blueslip.warn("User group with the given name does not exist.");
```
This is one thing I'm not too sure about how to handle. Should I use `warn` or `error` here?

timabbott replied:
error seems appropriate for a thing we think is impossible, since that will log to Sentry.

shubham-padia replied:
Fixed [here](https://github.com/zulip/zulip/compare/e009972cafbcd6f5ca3f3a6ec5ff7f9111f0b5bc..2da7989c19501ca1751ae6e8a34ef7fe3aa97992)

timabbott commented on web/src/user_groups.ts:
```diff
@@ -202,6 +202,23 @@ export function get_recursive_subgroups(target_user_group: UserGroup): Set<numbe
+    }
+
+    for (const subgroup_id of subgroup_ids) {
+        const subgroup = user_group_by_id_dict.get(subgroup_id)!;
```
An assert would probably be most appropriate for this scenario.

shubham-padia replied:
Fixed [here](https://github.com/zulip/zulip/compare/5a262b5339d4eb13585ab978b2caf053cf2e05d5..748e2d8c97be19d56506cd0057924da29a0ee134). But I'm curious to know why an `assert` statement over `!`, is it because ! checks for both null and undefined and we want to check only for undefined here?

timabbott commented on web/src/user_group_pill.ts:
```diff
@@ -17,8 +18,12 @@ export type UserGroupPillData = UserGroup & {
+    // In case of 0, which is technically possible in case all the
+    // users in a group are deactivated, we will show 0 users.
+    const user_pluralized_text = group_members.length === 1 ? "user" : "users";
+    return `${user_groups.get_display_group_name(group)}: ${group_members.length} ${user_pluralized_text}`;
```
This was a preexisting issue, but this is not valid internationalization. We need to use the ICU plurals syntax for this.

See `git grep ' plural, ' web/src/` and the discussion on plurals in https://zulip.readthedocs.io/en/latest/translating/internationalization.html for details.

shubham-padia replied:
Fixed in 213ce76f0598ef67ccf03f5b1cba21e690b3d313

timabbott commented:
This lgtm aside from the minor comments above.

shubham-padia commented:
Addressed the comments, and aside from the existing PR, added a test to make sure deactivated users are excluded from user_group_pill.get_user_ids which can be found in this [diff](https://github.com/zulip/zulip/compare/e009972cafbcd6f5ca3f3a6ec5ff7f9111f0b5bc..2da7989c19501ca1751ae6e8a34ef7fe3aa97992). 

This should be good to review now.

alya commented:
@timabbott Your feedback has been addressed; I haven't re-tested.

timabbott commented:
Merged, thanks @shubham-padia! Glad to have this checked off.
